go/parquet/file/column_writer.go

```diff
@@ -0,0 +1,567 @@
+
+	"github.com/apache/arrow/go/arrow"
+	"github.com/apache/arrow/go/arrow/array"
+	"github.com/apache/arrow/go/arrow/bitutil"
+	"github.com/apache/arrow/go/arrow/memory"
+	"github.com/apache/arrow/go/parquet"
+	"github.com/apache/arrow/go/parquet/internal/encoding"
+	"github.com/apache/arrow/go/parquet/metadata"
+	"github.com/apache/arrow/go/parquet/schema"
+)
+
+//go:generate go run ../../arrow/_tools/tmpl/main.go -i -data=../internal/encoding/physical_types.tmpldata column_writer_types.gen.go.tmpl
+
+// ColumnChunkWriter is the base interface for all columnwriters. To directly write
+// data to the column, you need to assert it to the correctly typed ColumnChunkWriter
+// instance, such as Int32ColumnWriter.
+type ColumnChunkWriter interface {
+	// Close ends this column and returns the number of bytes written
+	Close() error
+	// Type returns the underlying physical parquet type for this column
+	Type() parquet.Type
+	// Descr returns the column information for this writer
+	Descr() *schema.Column
+	// RowsWritten returns the number of rows that have so far been written with this writer
+	RowsWritten() int64
+	// TotalCompressedBytes returns the number of bytes, after compression, that have been written so far
+	TotalCompressedBytes() int64
+	// TotalBytesWritten includes the bytes for writing dictionary pages, while TotalCompressedBytes is
+	// just the data and page headers
+	TotalBytesWritten() int64
+	// Properties returns the current WriterProperties in use for this writer
+	Properties() *parquet.WriterProperties
+
+	LevelInfo() LevelInfo
+	SetBitsBuffer(*memory.Buffer)
+}
+
+func computeLevelInfo(descr *schema.Column) (info LevelInfo) {
+	info.DefLevel = descr.MaxDefinitionLevel()
+	info.RepLevel = descr.MaxRepetitionLevel()
+
+	minSpacedDefLevel := descr.MaxDefinitionLevel()
+	n := descr.SchemaNode()
+	for n != nil && n.RepetitionType() != parquet.Repetitions.Repeated {
+		if n.RepetitionType() == parquet.Repetitions.Optional {
+			minSpacedDefLevel--
+		}
+		n = n.Parent()
+	}
+	info.RepeatedAncestorDefLevel = minSpacedDefLevel
+	return
+}
+
+type columnWriter struct {
+	metaData *metadata.ColumnChunkMetaDataBuilder
+	descr    *schema.Column
+
+	// scratch buffer if validity bits need to be recalculated
+	bitsBuffer *memory.Buffer
+	levelInfo  LevelInfo
+	pager      PageWriter
+	hasDict    bool
+	encoding   parquet.Encoding
+	props      *parquet.WriterProperties
+	defEncoder encoding.LevelEncoder
+	repEncoder encoding.LevelEncoder
+	mem        memory.Allocator
+
+	pageStatistics  metadata.TypedStatistics
+	chunkStatistics metadata.TypedStatistics
+
+	// total number of values stored in the data page. this is the maximum
+	// of the number of encoded def levels or encoded values. for
+	// non-repeated, required columns, this is equal to the number of encoded
+	// values. For repeated or optional values, there may be fewer data values
+	// than levels, and this tells you how many encoded levels there are in that case
+	numBuffered int64
```

Comment:
numBufferedValues or numBufferedLevels might be better?

Reply:
done